src/datatypes/uncertainty.ts

```diff
@@ -11,7 +12,7 @@ export class Uncertainty {
 
   constructor(public low: any = null, public high?: any) {
     const gt = (a: any, b: any) => {
-      if (typeof a !== typeof b) {
+      if (typeof a !== typeof b || a?.constructor?.name !== b?.constructor?.name) {
```

Comment:
This alternate constructor name based check could have issues with minimization (weird one letter names that could overlap) or even CQL `Date` vs JS `Date`.

Reply:
@hossenlopp - I don't think this will be problem.  My understanding of the minimization problem is that it occurs when constructor.name is being compared to a static string (something like `a.constructor.name == 'MyClass'`.  When minimizing, the class name will be mangled but the static string won't, which leads to the mismatch.

In the case above, the comparison is not to a static string.  Mangled or not, the comparison would remain consistent.  Similarly, I think __instance in the normalization code is ok.

I'm not sure I've seen an overlap - is there a case I'm not thinking of?

Reply:
I think the overlap that @hossenlopp is concerned with is if the minimizer re-uses constructor names (e.g., gives two classes the same minimized name, like `A`).  Sometimes code obfuscaters might do this on purpose to make it more difficult to de-obfuscate the code.  I don't know if JS minifiers do it (intentionally or otherwise) -- but if they do, then this could cause issues.

Given that `Uncertainty` only supports a small number of types, you could just run through the different `isX` getters we added in order to avoid relying on constructor names.  E.g., 
```ts
if (
  typeof a !== typeof b ||
  a.isDate != b.isDate ||
  a.isDateTime != b.isDateTime ||
  a.isQuantity != b.isQuantity
) {
```

Reply:
Or, actually, might this be a simpler way that works more generally:
```ts
if (typeof a !== typeof b || a?.constructor !== b?.constructor) {
```
I think that's more reliable than just the constructor name, and it _should_ work as long as we aren't mixing two different versions of `cql-execution` at the same time -- but now that `cql-exec-fhir` uses `cql-execution` as a _peerDependency_, that really shouldn't happen.

Reply:
Checked in a commit going with the `if (typeof a !== typeof b || a?.constructor !== b?.constructor)` solution.  

I also added a test case that is unrealistic (even perverse...) that fails if using `if (typeof a !== typeof b)` or the first suggestion, but that the second solution properly distinguishes.

How about the __instance checks in normalization?  It would be an incredible edge case that two different types with the same properties were mangled in such a way that the constructor names were also the same.  Nonetheless, any suggestions?

Reply:
Cool.  Thanks, @birick1.  One quick note: I think the comment in that new test is just a little off though, as it still refers to `constructor?.name`.


> How about the __instance checks in normalization?

I agree that it seems unlikely for two different instances to have a constructor w/ the same name, same fields, and same values -- yet _not_ be equal.  That said, IIRC, you're already switching on type, so perhaps instead of doing something like: 
```ts
case Date:
  return Immutable.Map({
    epochMs: js.getTime(),
    __instance: js.constructor.name
  });
```
you could hard-code in an instance name:
```ts
case Date:
  return Immutable.Map({
    epochMs: js.getTime(),
    __instance: 'CQL.Date'
  });
```
Adding a prefix like `CQL.` would ensure that in the default case (or other less specific cases), if you need to use `constructor.name`, there's no chance of accidental overlap (like there could be w/ a JS Date vs CQL Date).

Reply:
Argh - wrote comment first and then the case.  Updated.

The fall through case is still problematic.  Admittedly, it's at the edge of the edge case now:

```
      default:
        return Immutable.Seq(js)
          .map((x: any) => this.toImmutableObjectKey(x))
          .toMap()
          .set('__instance', js.constructor?.name);
```

Reply:
Yeah, there is still probably a need for `constructor?.name` in some cases (like the fall-through), but if we can minimize that surface area, we're probably in good shape.  And like you said, chances of clash were already low anyway.

That said, now I'm wondering how records passed back from `cql-exec-fhir` look in this.  They will all fall through (since they're FHIR resources, not CQL types), and most of them will have `constructor.name == 'FHIRObject'`.  The type information (e.g., resource type, etc) is only stored as a non-enumerable property.  Are non-enumerable properties used in the normalization?

Reply:
Updated to fixed names in the normalization (except fall through).  Kept the namespacing as in suggestion - `CQL.` and `JS.`.

***

Interesting.   No - non-enumerable properties are not used in normalization.  It's possible, but probably not a good idea.

I did a lot of testing around the query.ts Distinct implementation, which does frequently operate on FHIRObject.  Since the enumerable properties of FHIRObject are accessed via functions (the `get` function on the interface), that means all properties of the fhir object, even null/undefined, are accessed and mapped into the normalized key.  So an equality conflict would only occur if two FHIR objects have the same structure and values on all enumerable properties (impossible?).  

The method getTypeInfo in cql-exec-fhir doesn't seem to be required by the engine api.  If it were, here's a possible solution (in normalization fall through)...

![image](https://user-images.githubusercontent.com/109760379/195672261-4b46efce-66f1-49c8-bd43-1cfd19390872.png)

Reply:
> So an equality conflict would only occur if two FHIR objects of different types have the same structure and values on all enumerable properties (impossible?).

At first I was going to say it is probably impossible, but then I remembered that _all FHIR types_ (not just resources) are wrapped in FHIRObjects and subject to being found in a list. One issue is that all primitives have exactly the same property names. For example, here's a little snippet from the FHIR 4.0.1 modelinfo file:
```xml
   <typeInfo baseType="FHIR.Element" namespace="FHIR" name="date" identifier="http://hl7.org/fhir/StructureDefinition/date" label="date" retrievable="false" xsi:type="ClassInfo">
      <element name="value" elementType="System.Date"/>
   </typeInfo>
   <typeInfo baseType="FHIR.Element" namespace="FHIR" name="dateTime" identifier="http://hl7.org/fhir/StructureDefinition/dateTime" label="dateTime" retrievable="false" xsi:type="ClassInfo">
      <element name="value" elementType="System.DateTime"/>
   </typeInfo>
   <typeInfo baseType="FHIR.Element" namespace="FHIR" name="decimal" identifier="http://hl7.org/fhir/StructureDefinition/decimal" label="decimal" retrievable="false" xsi:type="ClassInfo">
      <element name="value" elementType="System.Decimal"/>
   </typeInfo>
```
As you can see, `date`, `dateTime`, and `decimal` all have the same properties (`value` and the properties inherited from `Element`: `id` and `extension`).

Even more problematic (or at least just as problematic) are the elements that are basically just re-cast strings:
```xml
   <typeInfo baseType="FHIR.string" namespace="FHIR" name="code" identifier="http://hl7.org/fhir/StructureDefinition/code" label="code" retrievable="false" xsi:type="ClassInfo"/>
   <typeInfo baseType="FHIR.string" namespace="FHIR" name="id" identifier="http://hl7.org/fhir/StructureDefinition/id" label="id" retrievable="false" xsi:type="ClassInfo"/>
   <typeInfo baseType="FHIR.string" namespace="FHIR" name="markdown" identifier="http://hl7.org/fhir/StructureDefinition/markdown" label="markdown" retrievable="false" xsi:type="ClassInfo"/>
   <typeInfo baseType="FHIR.Element" namespace="FHIR" name="string" identifier="http://hl7.org/fhir/StructureDefinition/string" label="string" retrievable="false" xsi:type="ClassInfo">
      <element name="value" elementType="System.String"/>
   </typeInfo>
```
These have the same exactly properties _and types_ (or, in other words, the same exact signature). So a `code` with value `foo` is indistinguishable from an `id` (or a `markdown` or a `string`) with value `foo`. In this case, there is definitely a possibility of calling two things equal which are not equal.

Reply:
> The method getTypeInfo in cql-exec-fhir doesn't seem to be required by the engine api. If it were, here's a possible solution (in normalization fall through)...

You're right about `getTypeInfo` not being required by the engine API, but I think you may be on to something. Since CQL Execution 2.0.0, any data source that wants to support the `is` or `as` operator must implement the `_is` and `_typeHierarchy` functions.  This is described in the [CQL Execution 2.0.0 Release Notes](https://github.com/cqframework/cql-execution/releases/tag/v2.0.0).  So... for at least some data sources (like cql-exec-fhir), we could leverage the `_typeHierarchy` function.  That returns an array of types for which the first item is the object's direct type.  So if you detect the `_typeHierarchy` function, you could use it to set the `__instance` -- and that would get us around the primitive issue.